src/ucp/core/ucp_worker.c

```diff
@@ -385,6 +494,125 @@ void ucp_worker_progress(ucp_worker_h worker)
     ucs_assert(--worker->inprogress == 0);
 }
 
+ucs_status_t ucp_worker_get_efd(ucp_worker_h worker, int *fd)
+{
+    int res_fd, tl_fd;
+    ucs_status_t status;
+    uct_wakeup_h wakeup;
+    ucp_rsc_index_t tl_id;
+    ucp_context_h context = worker->context;
+
+    if (worker->wakeup.wakeup_efd != -1) {
+        *fd = worker->wakeup.wakeup_efd;
+        return UCS_OK;
+    }
+
+    res_fd = epoll_create(context->num_tls);
+    if (res_fd == -1) {
+        ucs_error("Failed to create epoll descriptor: %m");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    status = ucp_worker_wakeup_add_fd(res_fd, worker->wakeup.wakeup_pipe[0]);
+    if (status != UCS_OK) {
+        goto epoll_cleanup;
+    }
+
+    for (tl_id = 0; tl_id < context->num_tls; tl_id++) {
+        wakeup = worker->wakeup.iface_wakeups[tl_id];
+        if (wakeup != NULL) {
+            status = uct_wakeup_efd_get(wakeup, &tl_fd);
+            if (status != UCS_OK) {
+                goto epoll_cleanup;
+            }
+
+            status = ucp_worker_wakeup_add_fd(res_fd, tl_fd);
+            if (status != UCS_OK) {
+                goto epoll_cleanup;
+            }
+        }
+    }
+
+    worker->wakeup.wakeup_efd = res_fd;
+    *fd = res_fd;
+    return UCS_OK;
+
+epoll_cleanup:
+    close(res_fd);
+    return status;
+}
+
+ucs_status_t ucp_worker_arm(ucp_worker_h worker)
+{
+    int res;
+    char buf;
+    ucs_status_t status;
+    uct_wakeup_h wakeup;
+    ucp_rsc_index_t tl_id;
+    ucp_context_h context = worker->context;
+
+    for (tl_id = 0; tl_id < context->num_tls; ++tl_id) {
+        wakeup = worker->wakeup.iface_wakeups[tl_id];
+        if (wakeup != NULL) {
+            status = uct_wakeup_efd_arm(wakeup);
+            if (status != UCS_OK) {
+                return status;
+            }
+        }
+    }
+
+    do {
+        res = read(worker->wakeup.wakeup_pipe[0], &buf, 1);
+    } while (res != -1);
+
+    if (errno != EAGAIN) {
+        ucs_error("Read from internal pipe failed: %m");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucp_worker_wait(ucp_worker_h worker)
+{
+    int res;
+    int epoll_fd;
+    ucs_status_t status;
+    struct epoll_event *events;
+    ucp_context_h context = worker->context;
+
+    status = ucp_worker_get_efd(worker, &epoll_fd);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    events = alloca(context->num_tls * sizeof(*events));
```

Comment:
Using alloca making our future port for BSD systems is more cumbersome. If there is no real need, please avoid using it.